lib/Glimmer.pm

```diff
@@ -0,0 +1,494 @@
+    #-----------------------------------------------------------------------
+    my $tmp_predictions = "$tmp_prefix.predict";
+    print STDERR "\nParsing $tmp_predictions\n"
+	if $verbose;
+    
+    open(PREDICT,  "<$tmp_predictions")
+	|| die "Could not read-open $tmp_predictions";
+    
+    my $fid_num = 0;
+
+    my @output;
+
+    $contig_id  = qq();
+    while(defined($entry = <PREDICT>)) {
+	chomp $entry;
+
+	if ($entry =~ m/^>(\S+)/) {
+	    $contig_id = $1;
+	    next;
+	}
+	elsif ($entry =~ m/^(\S+)\s+(\d+)\s+(\d+)/) {
+	    my ($id, $beg, $end) = ($1, $2, $3);
+	    
+	    if ($beg && $end) {
+		++$fid_num;
+		my $fid = $id_prefix . $fid_num;
+		# print STDOUT (qq($fid\t), join(qq(_), ($contig_id, $beg, $end)), qq(\n));
+		my $dna = get_dna_seq($contig_id, $beg, $end, $len_of, $seq_of);
+		push(@output, [$fid, $contig_id, $beg, $end, $dna]);
+	    }
+	    else {
+		die "Error in $tmp_predictions line $.: $entry";
+	    }
+	}
+	else {
+	    die "Could not parse prediction $.: $entry";
+	}
+    }
+    
+    if (not $verbose) {
+	run($verbose, "rm -f $tmp_prefix.*");
+    }
+    else {
+	print STDERR "\nKeeping tmp files\n\n";
+    }
+    return \@output;
+}    
+    
+    ########################################################################
+sub get_dna_seq {
+    my ($contig, $beg, $end, $len_of, $seq_of) = @_;
+    my $dna = qq();
+    
+    my $contig_len = $len_of->{$contig};
+    my $contig_seq = $seq_of->{$contig};
+    
+    return undef if (&max($beg, $end) > $contig_len);
+    return undef if (&min($beg, $end) < 1);
+    
+    if ($beg < $end) {
+	$dna = substr($contig_seq, ($beg-1), ($end+1-$beg));
+    }
+    else {
+	$dna = substr($contig_seq, ($end-1), ($beg+1-$end));
+	$dna = $ { &rev_comp(\$dna) };
+    }
+    
+    return lc($dna);
+}
+
+
+sub read_fasta_record
+{
+    my ($file_handle) = @_;
+    my ( $old_end_of_record, $fasta_record, @lines, $head, $sequence, $seq_id, $comment, @parsed_fasta_record );
+    
+    if (not defined($file_handle))  { $file_handle = \*STDIN; }
+    
+    $old_end_of_record = $/;
+    $/ = "\n>";
+    
+    if (defined($fasta_record = <$file_handle>))
+    {
+	chomp $fasta_record;
+	@lines  =  split( /\n/, $fasta_record );
+	$head   =  shift @lines;
+	$head   =~ s/^>?//;
+	$head   =~ m/^(\S+)/;
+	$seq_id = $1;
+	
+	if ($head  =~ m/^\S+\s+(.*)$/)  { $comment = $1; } else { $comment = ""; }
+	
+	$sequence  =  join( "", @lines );
+	
+	@parsed_fasta_record = ( $seq_id, \$sequence, $comment );
+    }
+    else
+    {
+	@parsed_fasta_record = ();
+    }
+    
+    $/ = $old_end_of_record;
+    
+    return @parsed_fasta_record;
+}
+
+sub display_id_and_seq 
+{
+    my( $id, $seq, $fh ) = @_;
+    
+    if (! defined($fh) )  { $fh = \*STDOUT; }
+    
+    print $fh ">$id\n";
+    &display_seq($seq, $fh);
+}
+
+sub display_seq
+{
+    my ( $seq, $fh ) = @_;
+    my ( $i, $n, $ln );
+    
+    if (! defined($fh) )  { $fh = \*STDOUT; }
+    
+    $n = length($$seq);
```

Comment:
I would just pass in the sequence, rather than a reference to it -- unless there is a really compelling reason to pass around references instead of strings, I always pass scalars, rather than references to scalars, when writing perl.